src/Symfony/Bundle/DependencyInjection/Compiler/ValidationGroupsGeneratorPass.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Symfony\Bundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Injects validation groups generators.
+ *
+ * @internal
+ *
+ * @author Danny van Wijk <dannyvanwijk@gmail.com>
+ */
+final class ValidationGroupsGeneratorPass implements CompilerPassInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container): void
+    {
+        $generators = [];
+        foreach ($container->findTaggedServiceIds('api_platform.validation_groups_generator', true) as $serviceId => $tags) {
+            foreach ($tags as $tag) {
+                $generators[$tag['id'] ?? $serviceId] = new Reference($serviceId);
+            }
+        }
+
+        $container->getDefinition('api_platform.validator_locator')->addArgument($generators);
```

Comment:
Can't we try to use `tagged_locator` instead of a compiler pass, like suggested by @dunglas?

Reply:
That was my first implementation but didn't work because of this error "Failed asserting that exception message 'Invalid definition for service "api_platform.validator_locator": an array of references is expected as first argument when the "container.service_locator" tag is set.' contains '$json must be array or string (JSON array or JSON object)'."

Reply:
I don't think it is the way to do it.
You don't need to define a new service, just change the configuration of `api_platform.validator`:
```php
        <service id="api_platform.validator" class="ApiPlatform\Symfony\Validator\Validator">
            <argument type="service" id="validator" />
            <argument type="tagged_locator" tag="api_platform.validation_groups_generator" />
        </service>
```

Reply:
That works thanks, pr is updated.